src/librustc_metadata/cstore.rs

```diff
@@ -94,50 +95,48 @@ pub struct CrateMetadata {
     crate extern_crate: Lock<Option<ExternCrate>>,
 }
 
+#[derive(Clone)]
 pub struct CStore {
-    metas: RwLock<IndexVec<CrateNum, Option<Lrc<CrateMetadata>>>>,
-    crate metadata_loader: Box<dyn MetadataLoader + Sync>,
+    metas: IndexVec<CrateNum, Option<Lrc<CrateMetadata>>>,
```

Comment:
Unfortunately we cannot remove this `Lrc` because `CStore` has to be cloneable.

Reply:
Wow, I didn't even realize that was a thing. Do you know where the clones are coming from?

Reply:
I guess we could get rid of it by having the `expansion` query produce the `CStore` structure and redirect later uses to that instead. That would make the structure more like before this PR. It doesn't seem like a huge win right now though. It might be worth waiting on moving things into TyCtxt. Maybe we'd replace it by a `crate_metadata(CrateNum) -> &'tcx CrateMetadata` query or something.

Reply:
Yeah, this certainly can be done later if "later" is a better time due to other changes (like `TyCtxt`).
The clone itself (the primary effect) is cheap - only a vector of `Lrc`s is cloned.
The locks around `CrateMetadata::{dependencies,dep_kind,extern_crate}` (the secondary effect) are also unlikely to be accessed often and cause performance issues.